#!/bin/bash

#########################################################
#
# Platform: NCI Gadi HPC
# Description: see https://github.com/Sydney-Informatics-Hub/Shotgun-Metagenomics-Analysis
#
# Author/s: Cali Willet
# cali.willet@sydney.edu.au
#
# If you use this script towards a publication, please acknowledge the
# Sydney Informatics Hub (or co-authorship, where appropriate).
#
# Suggested acknowledgement:
# The authors acknowledge the scientific and technical assistance
# <or e.g. bioinformatics assistance of <PERSON>> of Sydney Informatics
# Hub and resources and services from the National Computational
# Infrastructure (NCI), which is supported by the Australian Government
# with access facilitated by the University of Sydney.
#
#########################################################

#PBS -P <project>
#PBS -N bb-prep
#PBS -l walltime=02:00:00
#PBS -l ncpus=48
#PBS -l mem=190GB
#PBS -q normal
#PBS -W umask=022
#PBS -l wd
#PBS -o ./Logs/bbmap_prep.o
#PBS -e ./Logs/bbmap_prep.e
#PBS -l storage=scratch/<project>


module load bbtools/37.98 # this is not a gloabl app - please self-install 

mkdir -p ./ref

reference=<reference> # full path and filename of your host species reference genome sequence (.fasta format)
prefix=$(basename $reference)

#repeat mask the reference:
bbmask.sh \
	in=${reference} \
	out=./ref/${prefix}-BBmasked \
	overwrite=t
		
#index the reference:
bbmap.sh -Xmx161g \
	ref=./ref/${prefix}-BBmasked



