#!/bin/bash

#########################################################
#
# Platform: NCI Gadi HPC
# Description: see https://github.com/Sydney-Informatics-Hub/Shotgun-Metagenomics-Analysis
#
# Author/s: Cali Willet
# cali.willet@sydney.edu.au
#
# If you use this script towards a publication, please acknowledge the
# Sydney Informatics Hub (or co-authorship, where appropriate).
#
# Suggested acknowledgement:
# The authors acknowledge the scientific and technical assistance
# <or e.g. bioinformatics assistance of <PERSON>> of Sydney Informatics
# Hub and resources and services from the National Computational
# Infrastructure (NCI), which is supported by the Australian Government
# with access facilitated by the University of Sydney.
#
#########################################################

#PBS -P <project>
#PBS -N bracken-reads
#PBS -l walltime=00:15:00
#PBS -l ncpus=1
#PBS -l mem=4GB
#PBS -q normal
#PBS -W umask=022
#PBS -l wd
#PBS -o ./Logs/bracken2_reads.o
#PBS -e ./Logs/bracken2_reads.e
#PBS -lstorage=scratch/<project>

module load python3/3.7.4 kraken2/2.0.8-beta bracken/2.6.0

src=$(which bracken)
src=${src/%bracken/src}

config=./Inputs/<cohort>.config

KRAKEN_DB=kraken2_standard_db_build_<date> ###UPDATE TO YOUR KRAKEN2 DB
KMER_LEN=35
READ_LEN=150
CLASSIFICATION_LVL=S #default=S (species)
THRESHOLD=10 #Default=10

outdir=./Abundance_reads
mkdir -p $outdir

kdir=./Speciation_reads

samples=$(awk 'NR>1 {print $2}' $config)
samples=($samples)

for (( i = 0; i < ${#samples[@]}; i++))
do
	sample=${samples[$i]}
	samplekdir=./${kdir}/${sample}
	kreport=${samplekdir}/${sample}.kraken2.standard.report

	sampleoutdir=./${outdir}/${sample}
	mkdir -p $sampleoutdir
	out=${sampleoutdir}/${sample}.bracken

	python3 ${src}/est_abundance.py \
		-i ${kreport} \
		-k ${KRAKEN_DB}/database${READ_LEN}mers.kmer_distrib \
		-l ${CLASSIFICATION_LVL} \
		-t ${THRESHOLD} \
		-o ${out}
done
