#!/bin/bash

#########################################################
#
# Platform: NCI Gadi HPC
# Description: see https://github.com/Sydney-Informatics-Hub/Shotgun-Metagenomics-Analysis
#
# Author: Tracy Chew
# tracy.chew@sydney.edu.au
#
# If you use this script towards a publication, please acknowledge the
# Sydney Informatics Hub (or co-authorship, where appropriate).
#
# Suggested acknowledgement:
# The authors acknowledge the scientific and technical assistance
# <or e.g. bioinformatics assistance of <PERSON>> of Sydney Informatics
# Hub and resources and services from the National Computational
# Infrastructure (NCI), which is supported by the Australian Government
# with access facilitated by the University of Sydney.
#
#########################################################

#PBS -P <project>
#PBS -l walltime=10:00:00
#PBS -l ncpus=28
#PBS -l mem=128GB
#PBS -q normalbw
#PBS -W umask=022
#PBS -l jobfs=400GB
#PBS -l wd
#PBS -l storage=scratch/<project>

set -e

module load diamond/2.0.11

# Change paths to your diamond makedb output and NCBI database:
database_dir=./NCBI_NR

nr=${database_dir}/nr.dmnd
taxonmap=${database_dir}/prot.accession2taxid.FULL.gz

contigs=./Prodigal_CDS/${sample}.CDS.prot.fa
outdir=./Diamond_NCBI
TMPDIR=${PBS_JOBFS}/tmp
PTMPDIR=${PBS_JOBFS}/ptmp

out=${outdir}/${sample}.DIAMOND.tsv

mkdir -p ${TMPDIR} ${PTMPDIR}

diamond blastp \
	--threads ${NCPUS} \
	--db ${nr} \
	--taxonmap ${taxonmap} \
	--query ${contigs} \
	--max-target-seqs 1 \
	--outfmt 6 qseqid sseqid pident length mismatch gapopen qstart qend sstart send evalue bitscore staxids sscinames sskingdoms skingdoms sphylums \
	--evalue 0.00001 \
	--id 75 \
	--query-cover 75 \
	--out ${out}
